### PR TITLE
Use pkg/forwarder for submitting process-agent payloads

### PR DESCRIPTION
**Pull request by kevinconaway**
### What does this PR do?

This commit modifies the process-agent to use the pkg/forwarder facility for sending HTTP payloads.  This will ensure that payloads are retried until they are successful.

One difference is that this will submit the API key in a different header than we are currently using.  The intake API will be modified to read this header.

We're also sending a new header, X-DD-Agent-Timestamp that indicates the time that we attempted to send the payload

### Motivation

What inspired you to submit this pull request?

### Additional Notes

QA:

To QA, we need to make sure that process, container, pod and connection messages are all correctly sent to intake with. and without an HTTP proxy

Additionally, we want make sure that real time mode is toggled on and off appropriately

### CODE REVIEW

**Comment by kevinconaway** [comment]
One other point to note here is that this will now use the main agent proxy configuration instead of the process agent proxy configuration.  My assumption was that this would not be a problem because if the process agent is proxied than the main agent would be as well, with the same configuration.

File: pkg/clusteragent/orchestrator/controller.go [hunk 1]
```diff
@@ -134,13 +151,41 @@ func (o *Controller) processPods() {
...
+		extraHeaders.Set(api.ClusterIDHeader, o.clusterID)
+
+		body, err := encodePayload(m)
+		if err != nil {
+			log.Errorf("Unable to encode message: %s", err)
+			continue
+		}
+
+		payloads := forwarder.Payloads{&body}
+		responses, err := o.forwarder.SubmitPodChecks(payloads, extraHeaders)
```
Comment (mfpierre) [inline pkg/clusteragent/orchestrator/controller.go]:
```suggestion
		_, err := o.forwarder.SubmitPodChecks(payloads, extraHeaders)
```
we can ignore responses here and remove the range below

File: cmd/process-agent/collector.go [hunk 2]
```diff
@@ -134,17 +147,47 @@ func (l *Collector) run(exit chan bool) {
...
+					case checks.RTProcess.Name():
+						responses, err = l.forwarder.SubmitRTProcessChecks(payloads, extraHeaders)
+					case checks.Container.Name():
+						responses, err = l.forwarder.SubmitContainerChecks(payloads, extraHeaders)
+					case checks.RTContainer.Name():
+						responses, err = l.forwarder.SubmitRTContainerChecks(payloads, extraHeaders)
+					case checks.Connections.Name():
+						responses, err = l.forwarder.SubmitConnectionChecks(payloads, extraHeaders)
+					case checks.Pod.Name():
+						responses, err = l.forwarder.SubmitPodChecks(payloads, extraHeaders)
```
Comment (mfpierre) [inline cmd/process-agent/collector.go]:
```suggestion
						responses, err = l.podForwarder.SubmitPodChecks(payloads, extraHeaders)
```

File: pkg/clusteragent/orchestrator/controller.go (hunk 1 above)
Comment (kevinconaway) [reply pkg/clusteragent/orchestrator/controller.go]: I had left that in as a precautionary measure in case the resulting channel was changed to be unbuffered or of a different size than it currently is.  As of now, the channel is sized to the number of output transactions so none of the worker threads will be blocked writing to it but if the channel size was changed, the worker threads could be blocked if nothing is consuming from the channel.

Thoughts?

Comment (mfpierre) [reply pkg/clusteragent/orchestrator/controller.go]: sounds good to me, let's keep it 👍

File: cmd/process-agent/collector.go [hunk 3]
```diff
@@ -67,7 +69,8 @@ func NewCollector(cfg *config.AgentConfig) (Collector, error) {
 		rtIntervalCh:  make(chan time.Duration),
 		cfg:           cfg,
 		groupID:       rand.Int31(),
-		apiClient:     api.NewClient(http.Client{Timeout: HTTPTimeout, Transport: cfg.Transport}, ReqCtxTimeout),
+		forwarder:     forwarder.NewDefaultForwarder(keysPerDomains(cfg.APIEndpoints)),
```
Comment (sunhay) [inline cmd/process-agent/collector.go]: This is a convenient way of setting things up :)

Comment (sunhay) [reply cmd/process-agent/collector.go]: Does the `DefaultForwarder` setup it's own `Transport` using the config? We'll want to make sure it does so that proxy settings continue to work

Comment (kevinconaway) [reply cmd/process-agent/collector.go]: Yes.  Per my comment above, it uses the proxy settings from the main agent config though.

My assumption was that if proxying was configured on the process agent then the same configuration would be applied to the main agent.

**Comment by leeavital** [comment]
Please add QA steps in the PR description we can use during the release process

File: cmd/process-agent/collector_api_test.go [hunk 4]
```diff
@@ -19,56 +22,222 @@ import (
...
+
+		reqBody, err := process.DecodeMessage(req.body)
+		require.NoError(t, err)
+
+		_, ok := reqBody.Body.(*process.CollectorRealTime)
+		require.True(t, ok)
+
+		assert.Equal(t, "1", req.headers.Get("X-DD-Agent-Attempts"))
+
+		select {
```
Comment (leeavital) [inline cmd/process-agent/collector_api_test.go]: could we have runCollectorTest read until `payloads` is closed and then flush to get rid of the need to sleep here?

Comment (kevinconaway) [reply cmd/process-agent/collector_api_test.go]: The sleep isn't ideal but I'm not aware a better way to test that something _doesn't_ happen, when accounting for an asynchronous i/o call like this.

We don't have access to the `payloads` channel here unless I'm misunderstanding what you're suggesting.

File: pkg/clusteragent/orchestrator/controller.go [hunk 5]
```diff
@@ -8,11 +8,16 @@
...
 import (
+	"fmt"
 	"math/rand"
 	"net/http"
 	"sync/atomic"
 	"time"
 
+	model "github.com/DataDog/agent-payload/process"
+
+	"github.com/DataDog/datadog-agent/pkg/forwarder"
```
Comment (mfpierre) [inline pkg/clusteragent/orchestrator/controller.go:19]: nit: group this up with import below

**Review by mfpierre — DISMISSED** [review]
Looking good on the container-app side, nice rework 🙌

File: pkg/forwarder/forwarder.go [hunk 6]
```diff
@@ -346,3 +383,87 @@ func (f *DefaultForwarder) SubmitV1Intake(payload Payloads, extra http.Header) e
...
+
+	return f.submitProcessLikePayload(podEndpoint, payload, extra, true)
+}
+
+func (f *DefaultForwarder) submitProcessLikePayload(ep endpoint, payload Payloads, extra http.Header, retryable bool) (chan Response, error) {
+	transactions := f.createHTTPTransactions(ep, payload, false, extra)
+
+	results := make(chan Response, len(transactions))
+
+	var wg sync.WaitGroup
```
Comment (shang-wang) [inline pkg/forwarder/forwarder.go:434]: What's async here that a wg is needed?

File: pkg/clusteragent/orchestrator/controller.go (hunk 5 above)
Comment (kevinconaway) [reply pkg/clusteragent/orchestrator/controller.go:19]: Hmm, this is how go imports formats them for me.   Does it do something different for you?

File: pkg/forwarder/forwarder.go (hunk 6 above)
Comment (kevinconaway) [reply pkg/forwarder/forwarder.go:434]: Submitting the "http transaction" to the workers.  We will create txn per (domain,  apiKey) combination and each will be submitted. separately.

When all. the requests complete, the results channel will be closed (via a separate goroutine waiting on the wait group)

Comment (shang-wang) [reply pkg/forwarder/forwarder.go:434]: Ah got it.

File: pkg/clusteragent/orchestrator/controller.go [hunk 7]
```diff
@@ -107,13 +118,19 @@ func (o *Controller) Run(stopCh <-chan struct{}) {
 	log.Infof("Starting orchestrator controller")
 	defer log.Infof("Stopping orchestrator controller")
 
+	if err := o.forwarder.Start(); err != nil {
+		log.Errorf("error starting pod forwarder: %s", err)
```
Comment (shang-wang) [inline pkg/clusteragent/orchestrator/controller.go:122]: If forwarder didn't start we don't need to return?

File: pkg/clusteragent/orchestrator/controller.go (hunk 5 above)
Comment (mfpierre) [reply pkg/clusteragent/orchestrator/controller.go:19]: in the agent we usually have 3 groups for imports
- native
- project
- 3rd party

IIRC, if you group them up with the ones below they should stick to the group and be ordered alphabetically

File: cmd/process-agent/collector.go (hunk 3 above)
Comment (sunhay) [reply cmd/process-agent/collector.go:72]: It is - this is great :)

File: cmd/process-agent/collector.go [hunk 8]
```diff
@@ -118,6 +121,16 @@ func (l *Collector) run(exit chan bool) {
 	}
 	log.Infof("Starting process-agent for host=%s, endpoints=%s, orchestrator endpoints=%s, enabled checks=%v", l.cfg.HostName, eps, orchestratorEps, l.cfg.EnabledChecks)
 
+	if err := l.forwarder.Start(); err != nil {
+		log.Errorf("Error starting forwarder: %s", err)
+		return
+	}
+
+	if err := l.podForwarder.Start(); err != nil {
```
Comment (sunhay) [inline cmd/process-agent/collector.go:129]: Should we only do this if the `orchestrator` check is enabled?

File: pkg/forwarder/transaction.go [hunk 9]
```diff
@@ -210,7 +210,7 @@ func (t *HTTPTransaction) internalProcess(ctx context.Context, client *http.Clie
 		tlmTxErrors.Inc(t.Domain, "cant_send")
 		return 0, nil, fmt.Errorf("error while sending transaction, rescheduling it: %s", httputils.SanitizeURL(err.Error()))
 	}
-	defer resp.Body.Close()
+	defer func() { _ = resp.Body.Close() }()
```
Comment (sunhay) [inline pkg/forwarder/transaction.go:213]: Nice catch - this means there was a leak beforehand, no?

Comment (kevinconaway) [reply pkg/forwarder/transaction.go:213]: Nope, this is just removing an IDE warning about ignoring an error response.

File: cmd/process-agent/collector.go (hunk 8 above)
Comment (kevinconaway) [reply cmd/process-agent/collector.go:129]: I guess we could, we would have to surround this and the stop below with an if check.  Is it worth it to avoid having idle goroutines sitting around?

**Review by leeavital — DISMISSED** [review]

File: cmd/process-agent/collector.go [hunk 10]
```diff
@@ -258,3 +289,63 @@ func getContainerCount(mb model.MessageBody) int {
...
+	for response := range responses {
+		if response.Err != nil {
+			log.Errorf("Error from %s: %s", response.Domain, response.Err)
+			continue
+		}
+
+		if response.StatusCode >= 300 {
+			log.Errorf("Invalid response from %s: %d -> %s", response.Domain, response.StatusCode, response.Err)
+			continue
+		}
```
Comment (prognant) [inline cmd/process-agent/collector.go:305]: Are we 100% sure that 3xx / redirect code should be considered as invalid ?

**Review by prognant — DISMISSED** [review]

File: pkg/forwarder/transaction.go (hunk 9 above)
Comment (hush-hush) [inline pkg/forwarder/transaction.go:213]: Why this change ?

File: pkg/forwarder/forwarder.go [hunk 11]
```diff
@@ -346,3 +383,87 @@ func (f *DefaultForwarder) SubmitV1Intake(payload Payloads, extra http.Header) e
...
+				Err:        err,
+			}
+			wg.Done()
+		}
+	}
+
+	go func() {
+		wg.Wait()
+		close(results)
+	}()
```
Comment (hush-hush) [inline pkg/forwarder/forwarder.go:466]: This will create a goroutine by transaction currently being queued. We might drop a transaction without calling the completion handler for a number of reasons ([retry queue is fulled](https://github.com/DataDog/datadog-agent/blob/master/pkg/forwarder/domain_forwarder.go#L108), the domain was black listed, ...).

I don't think we should rely on it being closed or add a dropped handler and call it from everywhere. I'm less in favor of the second as it's so easy to forget to call it in future development leading to a major leak in the agent.

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: pkg/forwarder/transaction.go (hunk 9 above)
Comment (kevinconaway) [reply pkg/forwarder/transaction.go:213]: This removes a warning about ignoring an error being returned (`resp.Body.Close()` returns an error)

File: cmd/process-agent/collector.go (hunk 10 above)
Comment (kevinconaway) [reply cmd/process-agent/collector.go:305]: Yes, this would normally be handled inside the `net/http`.  Anything that makes its way back to the caller that isn't 2xx is invalid.

File: pkg/forwarder/forwarder.go (hunk 11 above)
Comment (kevinconaway) [reply pkg/forwarder/forwarder.go:466]: Good point, I will rewrite this so that there is not a reliance on the channel being closed.

**Review by mfpierre — DISMISSED** [review]

File: pkg/forwarder/forwarder.go [hunk 12]
```diff
@@ -346,3 +387,98 @@ func (f *DefaultForwarder) SubmitV1Intake(payload Payloads, extra http.Header) e
...
+
+	go func() {
+		receivedResponses := 0
+		for {
+			select {
+			case r := <-internalResults:
+				results <- r
+				receivedResponses++
+				if receivedResponses == expectedResponses {
+					close(results)
```
Comment (hush-hush) [inline pkg/forwarder/forwarder.go:473]: Should we not return where ? Seems like we're going to run the loop until the timeout otherwise and double close a channel (which will panic).

**Review by hush-hush — CHANGES_REQUESTED** [review]

**Review by hush-hush — APPROVED** [review]

**Review by ahmed-mez — APPROVED** [review]
Reviewed only the changes in `pkg/clusteragent` owned by the team, LGTM 👍

**Review by shang-wang — APPROVED** [review]

**Review by mfpierre — APPROVED** [review]
